### ISSUE TITLE
refinement in a-arrow is not run

### ISSUE THREAD

**Issue by sorawee** [brownplt/pyret-lang#628]
```
fun test(x):
  false
end

a :: List<Number%(test)> = [list: 1, 2, 3]
```

We should either run the refinement, or consider this an invalid syntax.

**Comment by jpolitz** [comment]
I don't see an `a-arrow` here.  Do you mean in an `a-app`?  It's documented that things in type instantiation position aren't checked dynamically.

And it's hard to avoid e.g.

```
type Odd = Number%(is-odd)
a :: List<T> = [list: 1, 2, 3]
```

We could make this be an error at some well-formedness pass that has information about which types contain refinements.  That's a pretty confusing error at that point, and I feel like if it's already clear that we aren't checking for `Number`-ness, it's not that much of a leap to also not call the refinement.

Agree that this isn't great, though.

**Comment by sorawee** [comment]
> I feel like if it's already clear that we aren't checking for Number-ness, it's not that much of a leap to also not call the refinement.

Fair enough, but eventually we will check that, right? So I will leave this issue open until then.

By the way, It's probably equally confusing to see refinement in the following works, while the refinement in the first comment doesn't work.

```
fun test(x):
  false
end

a :: {
  f :: Number%(test)
} = {f: 1}
```

**Comment by blerner** [comment]
It's not guaranteed that we will, and likely that we won't...at least not dynamically.  Consider a lazy list
```
data LazyList<T>:
  | l-empty
  | l-cons(l-first :: () -> T, l-rest :: () -> LazyList<T>)
end
```

If you instantiate `LazyList<Number%(is-even)>`, there's no possible way to check that it's true without forcing the lazy list.  And since lazy lists can easily build cyclic or infinite structures...that would be bad.

In general, we don't know how to reliably, safely, efficiently, correctly, etc. check the use of a type parameter in a generic type, other than by statically approximating it with a static type checker.

**Comment by jpolitz** [comment]
@blerner I think the `LazyList` example is orthogonal, because it holds for any type that gets delayed by a function.  Pyret similarly wouldn't force the following and it doesn't use type instantiation or refinements:

```
data LazyNumList:
  | l-empty
  | l-cons(l-first :: ( -> Number), l-rest :: ( -> LazyNumList))
end

lln = l-cons(lam(): "not-a-num" end, lam(): l-empty end)
```

**Comment by blerner** [comment]
Good point :)  So, all the moreso that we don't check refinements in all cases, if we can't/won't check un-refined types in all cases!
